eks_infrastructure/upgrade_operation.sh

```diff
@@ -56,11 +57,19 @@ function upgrade_nodegroups() {
         --cluster ${CLUSTER} \
         --kubernetes-version ${EKS_VERSION} \
         --timeout 90m \
-        --region ${REGION}
+        --region ${REGION} || { UPGRADE_FAILED=1; FAILED_NODE_GROUPS+=( ${NODEGROUP} ); }
     done
   else
     echo "No Nodegroups present in the EKS cluster ${1}"
   fi
+
+  if [ ${UPGRADE_FAILED:-0} -eq 1 ]; then
```

Comment:
nit: you should just be able to check for whether or not the failed_node_groups array is empty or not, rather than also keeping track of upgrade failed and looping thru to print out results again

example

```
if [ ${#FAILED_NODE_GROUPS[@]} -ne 0 ]; then
    echo "The following node groups failed to upgrade"
    echo "${FAILED_NODE_GROUPS[@]}"
    exit 1
fi
```